README
This repository contains all of the platform-specific code required to run the GPII Personalization Framework on Linux with GNOME.

The following components can be found in the repository:

* usbDriveListener: a UserListener implementation that will detect a USB drive with an anonymous GPII user token installed on it
* gSettingsBridge: a Node.js module for accessing the gSettings API via JavaScript within a Node application


Changelog
2011-12-18  rousseau

	* [r6150] README: release 1.4.18
	* [r6149] scriptor: Display lines of 16 bytes instead of 24
	* [r6148] gscriptor: Display bytes of value 0x20 as ' ' instead of
	  '.'
	  
	  Thanks to Alain SAURAT for the patch
	* [r6147] gscriptor: Display hex dumps in lines of 16 bytes instead
	  of 17
	  
	  Thanks to Alain SAURAT for the patch
	* [r6146] smartcard_list.txt: 3 new ATRs

2011-12-15  rousseau

	* [r6140] smartcard_list.txt: 1 new ATR + generalisation of the
	  Reinet SCT WOK card

2011-12-14  rousseau

	* [r6138] smartcard_list.txt: 1 new ATR

2011-12-10  rousseau

	* [r6137] smartcard_list.txt: 1 new ATR

2011-12-07  rousseau

	* [r6136] smartcard_list.txt: 4 new ATRs

2011-12-01  rousseau

	* [r6127] smartcard_list.txt: 1 new ATR

2011-11-29  rousseau

	* [r6121] smartcard_list.txt: 2 new ATRs

2011-11-25  rousseau

	* [r6116] sort_smartcard_list.py: Do not mix spaces and tabs
	* [r6115] smartcard_list.txt: 3 new ATRs and reorder

2011-11-23  rousseau

	* [r6114] smartcard_list.txt: 1 new ATR

2011-11-15  rousseau

	* [r6106] smartcard_list.txt: 8 new ATRs

2011-11-06  rousseau

	* [r6101] smartcard_list.txt: 3 new ATRs

2011-10-26  rousseau

	* [r6084] smartcard_list.txt: 2 new ATRs

2011-10-23  rousseau

	* [r6068] smartcard_list.txt: 2 new ATRs

2011-10-18  rousseau

	* [r6048] smartcard_list.txt: Fix description
	* [r6047] smartcard_list.txt: 3 new ATRs

2011-10-16  rousseau

	* [r6046] smartcard_list.txt: 1 new ATR

2011-10-14  rousseau

	* [r6039] smartcard_list.txt: 1 new ATR

2011-10-11  rousseau

	* [r6031] smartcard_list.txt: Add card description
	* [r6030] smartcard_list.txt: 2 new ATRs

2011-10-06  rousseau

	* [r6004] smartcard_list.txt: 2 new ATRs

2011-10-04  rousseau

	* [r5998] smartcard_list.txt: update URL in card description

2011-09-24  rousseau

	* [r5966] smartcard_list.txt: 1 new ATR

2011-09-20  rousseau

	* [r5958] smartcard_list.txt: 1 new ATR

2011-09-17  rousseau

	* [r5957] smartcard_list.txt: 2 new ATRs

2011-09-16  rousseau

	* [r5956] smartcard_list.txt: 1 new ATR

2011-09-09  rousseau

	* [r5950] smartcard_list.txt: 11 ATRs

2011-08-29  rousseau

	* [r5920] smartcard_list.txt: 1 new ATR

2011-08-28  rousseau

	* [r5918] smartcard_list.txt: 1 new ATR

2011-08-27  rousseau

	* [r5917] smartcard_list.txt: 1 new ATR
	* [r5913] smartcard_list.txt: 3 new ATRs

2011-08-19  rousseau

	* [r5892] pcsc_scan.c: Remove extra tab

2011-08-17  rousseau

	* [r5889] smartcard_list.txt: 2 new ATRs

2011-08-09  rousseau

	* [r5888] smartcard_list.txt: 1 new ATR

2011-08-08  rousseau

	* [r5884] smartcard_list.txt: 1 new ATR
	* [r5883] smartcard_list.txt: 1 new ATR

2011-08-03  rousseau

	* [r5876] smartcard_list.txt: 3 new ATRs

2011-07-17  rousseau

	* [r5872] smartcard_list.txt: 2 new ATRs

2011-07-05  rousseau

	* [r5823] smartcard_list.txt: 3 new ATRs

2011-07-02  rousseau

	* [r5820] smartcard_list.txt: 1 new ATR

2011-06-30  rousseau

	* [r5817] smartcard_list.txt: 1 new ATR

2011-06-16  rousseau

	* [r5798] smartcard_list.txt: 1 new ATR

2011-06-15  rousseau

	* [r5792] smartcard_list.txt: Update copyright date
	* [r5791] smartcard_list.txt: 1 new ATR

2011-06-03  rousseau

	* [r5777] smartcard_list.txt: 1 new ATR

2011-05-28  rousseau

	* [r5768] smartcard_list.txt: 3 new ATRs

2011-05-27  rousseau

	* [r5767] smartcard_list.txt: 2 new ATRs

2011-05-17  rousseau

	* [r5760] smartcard_list.txt: 2 new ATRs

2011-05-14  rousseau

	* [r5748] smartcard_list.txt: 3 new ATRs

2011-05-07  rousseau

	* [r5717] smartcard_list.txt: 5 new ATRs

2011-04-21  rousseau

	* [r5702] smartcard_list.txt: 1 new ATR

2011-04-19  rousseau

	* [r5701] smartcard_list.txt: 2 new ATRs

2011-04-17  rousseau

	* [r5700] smartcard_list.txt: 1 new ATR

2011-04-12  rousseau

	* [r5696] smartcard_list.txt: 1 new ATR

2011-03-30  rousseau

	* [r5680] smartcard_list.txt: 3 new ATRs

2011-03-25  rousseau

	* [r5670] smartcard_list.txt: fix an ATR
	* [r5669] smartcard_list.txt: 4 new ATRs

2011-03-22  rousseau

	* [r5665] smartcard_list.txt: 1 new ATR

2011-03-21  rousseau

	* [r5664] smartcard_list.txt: 2 new ATRs

2011-03-18  rousseau

	* [r5661] smartcard_list.txt: 10 new ATRs

2011-03-09  rousseau

	* [r5648] smartcard_list.txt: 1 new ATR

2011-03-05  rousseau

	* [r5642] smartcard_list.txt: 4 new ATRs

2011-02-25  rousseau

	* [r5641] smartcard_list.txt: 1 new ATR

2011-02-24  rousseau

	* [r5635] smartcard_list.txt: 1 new ATR

2011-02-23  rousseau

	* [r5631] smartcard_list.txt: 1 new ATR
	* [r5630] smartcard_list.txt: 1 new ATR

2011-02-18  rousseau

	* [r5619] smartcard_list.txt: 2 new ATRs

2011-02-17  rousseau

	* [r5618] smartcard_list.txt: 2 new ATRs

2011-02-16  rousseau

	* [r5613] smartcard_list.txt: 1 new ATR

2011-02-15  rousseau

	* [r5598] smartcard_list.txt: 1 new ATR

2011-02-11  rousseau

	* [r5591] smartcard_list.txt: 1 new ATR

2011-02-10  rousseau

	* [r5590] smartcard_list.txt: 5 new ATRs

2011-02-09  rousseau

	* [r5588] smartcard_list.txt: 1 new ATR

2011-02-08  rousseau

	* [r5585] smartcard_list.txt: 4 new ATRs

2011-02-04  rousseau

	* [r5577] smartcard_list.txt: 2 new ATRs

2011-02-02  rousseau

	* [r5576] smartcard_list.txt: 1 new ATR

2011-01-29  rousseau

	* [r5571] smartcard_list.txt: 1 new ATR
	* [r5564] smartcard_list.txt: 5 new ATRs

2011-01-28  rousseau

	* [r5561] smartcard_list.txt: 2 new ATRs

2011-01-26  rousseau

	* [r5553] smartcard_list.txt: 1 new ATR

2011-01-25  rousseau

	* [r5552] smartcard_list.txt: 1 new ATR
	* [r5551] smartcard_list.txt: 1 new ATR

2011-01-19  rousseau

	* [r5529] smartcard_list.txt: 1 new ATR
	* [r5526] smartcard_list.txt: 2 new ATRs

2011-01-17  rousseau

	* [r5512] smartcard_list.txt: 1 new ATR

2011-01-14  rousseau

	* [r5511] smartcard_list.txt: 1 new ATR
	* [r5510] smartcard_list.txt: update 1 ATR description

2011-01-13  rousseau

	* [r5503] smartcard_list.txt: 1 new ATR
	* [r5502] smartcard_list.txt: 1 new ATR

2011-01-08  rousseau

	* [r5497] smartcard_list.txt: 1 new ATR

2011-01-06  rousseau


pcsc_scan.c
/*
    Scan and print all the PC/SC devices available
    Copyright (C) 2001-2011  Ludovic Rousseau <ludovic.rousseau@free.fr>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
*/

/* $Id: pcsc_scan.c 5892 2011-08-19 09:32:46Z rousseau $ */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>
#include <string.h>

#ifdef __APPLE__
#include <PCSC/wintypes.h>
#include <PCSC/winscard.h>
#else
#include <winscard.h>
#endif

#ifndef TRUE
#define TRUE 1
#define FALSE 0
#endif

#define TIMEOUT 1000	/* 1 second timeout */

/* command used to parse (on screen) the ATR */
#define ATR_PARSER "ATR_analysis"

#ifndef SCARD_E_NO_READERS_AVAILABLE
#define SCARD_E_NO_READERS_AVAILABLE 0x8010002E
#endif

int num_gpii_tokens = 3;
char *gpii_atrs[] = {
  "3B 8F 80 01 80 4F 0C A0 00 00 03 06 03 00 01 00 00 00 00 6A",
  "3B 8F 80 01 80 4F 0C A0 00 00 03 06 03 00 26 00 00 00 00 4D",
  "3B 8F 80 01 80 4F 0C A0 00 00 03 06 03 00 03 00 00 00 00 68"
};
char *gpii_tokens[] = {
  "intergalacticSheepherder",
  "alanMathisonTuring",
  "donaldKnuth"
};


#define test_rv(fct, rv, hContext) \
do { \
	if (rv != SCARD_S_SUCCESS) \
	{ \
		printf("%s%s: %s%s\n", red, fct, pcsc_stringify_error(rv), color_end); \
		(void)SCardReleaseContext(hContext); \
		exit(-1); \
	} \
} while(0)

void usage(void)
{
	printf("usage: pcsc_scan [-n] [-V] [-h]\n");
	printf("  -n : no ATR analysis\n");
	printf("  -V : print version number\n");
	printf("  -h : this help\n");
} /* usage */

int main(int argc, char *argv[])
{
	int current_reader;
	LONG rv;
	SCARDCONTEXT hContext;
	SCARD_READERSTATE *rgReaderStates_t = NULL;
	SCARD_READERSTATE rgReaderStates[1];
	DWORD dwReaders = 0, dwReadersOld;
	DWORD timeout;
	LPSTR mszReaders = NULL;
	char *ptr, **readers = NULL;
	int nbReaders, i;
	char atr[MAX_ATR_SIZE*3+1];	/* ATR in ASCII */
        char atr_packed[MAX_ATR_SIZE*3+1];
	char atr_command[sizeof(atr)+sizeof(ATR_PARSER)+2+1];
	int opt;
	int analyse_atr = TRUE;
	char *blue = "";
	char *red = "";
	char *magenta = "";
	char *color_end = "";
	int pnp = TRUE;
        int logged_in = FALSE;
        time_t logged_in_time;

	printf("PC/SC device scanner\n");
	printf("V " VERSION " (c) 2001-2011, Ludovic Rousseau <ludovic.rousseau@free.fr>\n");
	printf("Compiled with PC/SC lite version: " PCSCLITE_VERSION_NUMBER "\n");

	while ((opt = getopt(argc, argv, "Vhn")) != EOF)
	{
		switch (opt)
		{
			case 'n':
				analyse_atr = FALSE;
				break;

			case 'V':
				/* the version number is printed by default */
				return 1;
				break;

			case 'h':
				default:
				usage();
				return 1;
				break;
		}
	}

	if (argc - optind != 0)
	{
		usage();
		return 1;
	}

	/* check if terminal supports color */
	{
		const char *terms[] = { "linux", "xterm", "xterm-color", "Eterm", "rxvt", "rxvt-unicode" };
		char *term;

		term = getenv("TERM");
		if (term)
		{
			int i;

			/* for each known color terminal */
			for (i = 0; i < sizeof(terms) / sizeof(terms[0]); i++)
			{
				/* we found a supported term? */
				if (0 == strcmp(terms[i], term))
				{
					blue = "\33[34m";
					red = "\33[31m";
					magenta = "\33[35m";
					color_end = "\33[0m";
					break;
				}
			}
		}
	}

	rv = SCardEstablishContext(SCARD_SCOPE_SYSTEM, NULL, NULL, &hContext);
	test_rv("SCardEstablishContext", rv, hContext);

	rgReaderStates[0].szReader = "\\\\?PnP?\\Notification";
	rgReaderStates[0].dwCurrentState = SCARD_STATE_UNAWARE;

	rv = SCardGetStatusChange(hContext, 0, rgReaderStates, 1);
	if (rgReaderStates[0].dwEventState & SCARD_STATE_UNKNOWN)
	{
		timeout = TIMEOUT;
		printf("%sPlug'n play reader name not supported. Using polling every %ld ms.%s\n", magenta, timeout, color_end);
		pnp = FALSE;
	}
	else
	{
		timeout = INFINITE;
		printf("%sUsing reader plug'n play mechanism%s\n", magenta, color_end);
	}

get_readers:
	/* free memory possibly allocated in a previous loop */
	if (NULL != readers)
	{
		free(readers);
		readers = NULL;
	}

	if (NULL != rgReaderStates_t)
	{
		free(rgReaderStates_t);
		rgReaderStates_t = NULL;
	}

	/* Retrieve the available readers list.
	 *
	 * 1. Call with a null buffer to get the number of bytes to allocate
	 * 2. malloc the necessary storage
	 * 3. call with the real allocated buffer
	 */
	printf("%sScanning present readers...%s\n", red, color_end);
	rv = SCardListReaders(hContext, NULL, NULL, &dwReaders);
	if (rv != SCARD_E_NO_READERS_AVAILABLE)
		test_rv("SCardListReaders", rv, hContext);

	dwReadersOld = dwReaders;

	/* if non NULL we came back so free first */
	if (mszReaders)
	{
		free(mszReaders);
		mszReaders = NULL;
	}

	mszReaders = malloc(sizeof(char)*dwReaders);
	if (mszReaders == NULL)
	{
		printf("malloc: not enough memory\n");
		return 1;
	}

	*mszReaders = '\0';
	rv = SCardListReaders(hContext, NULL, mszReaders, &dwReaders);

	/* Extract readers from the null separated string and get the total
	 * number of readers */
	nbReaders = 0;
	ptr = mszReaders;
	while (*ptr != '\0')
	{
		ptr += strlen(ptr)+1;
		nbReaders++;
	}

	if (SCARD_E_NO_READERS_AVAILABLE == rv || 0 == nbReaders)
	{
		printf("%sWaiting for the first reader...%s", red, color_end);
		fflush(stdout);

		if (pnp)
		{
			rgReaderStates[0].szReader = "\\\\?PnP?\\Notification";
			rgReaderStates[0].dwCurrentState = SCARD_STATE_UNAWARE;

			rv = SCardGetStatusChange(hContext, INFINITE, rgReaderStates, 1);
			test_rv("SCardGetStatusChange", rv, hContext);
		}
		else
		{
			rv = SCARD_S_SUCCESS;
			while ((SCARD_S_SUCCESS == rv) && (dwReaders == dwReadersOld))
			{
				rv = SCardListReaders(hContext, NULL, NULL, &dwReaders);
				if (SCARD_E_NO_READERS_AVAILABLE == rv)
					rv = SCARD_S_SUCCESS;
				sleep(1);
			}
		}
		printf("found one\n");
		goto get_readers;
	}
	else
		test_rv("SCardListReader", rv, hContext);

	/* allocate the readers table */
	readers = calloc(nbReaders+1, sizeof(char *));
	if (NULL == readers)
	{
		printf("Not enough memory for readers table\n");
		return -1;
	}

	/* fill the readers table */
	nbReaders = 0;
	ptr = mszReaders;
	while (*ptr != '\0')
	{
		printf("%s%d: %s%s\n", blue, nbReaders, ptr, color_end);
		readers[nbReaders] = ptr;
		ptr += strlen(ptr)+1;
		nbReaders++;
	}

	/* allocate the ReaderStates table */
	rgReaderStates_t = calloc(nbReaders+1, sizeof(* rgReaderStates_t));
	if (NULL == rgReaderStates_t)
	{
		printf("Not enough memory for readers states\n");
		return -1;
	}

	/* Set the initial states to something we do not know
	 * The loop below will include this state to the dwCurrentState
	 */
	for (i=0; i<nbReaders; i++)
	{
		rgReaderStates_t[i].szReader = readers[i];
		rgReaderStates_t[i].dwCurrentState = SCARD_STATE_UNAWARE;
	}
	rgReaderStates_t[nbReaders].szReader = "\\\\?PnP?\\Notification";
	rgReaderStates_t[nbReaders].dwCurrentState = SCARD_STATE_UNAWARE;

	/* Wait endlessly for all events in the list of readers
	 * We only stop in case of an error
	 */
	if (pnp)
		nbReaders++;

	rv = SCardGetStatusChange(hContext, timeout, rgReaderStates_t, nbReaders);
	while ((rv == SCARD_S_SUCCESS) || (rv == SCARD_E_TIMEOUT))
	{
		time_t t;

		if (pnp)
		{
			if (rgReaderStates_t[nbReaders-1].dwEventState &
					SCARD_STATE_CHANGED)
				goto get_readers;
		}
		else
		{
			/* A new reader appeared? */
			if ((SCardListReaders(hContext, NULL, NULL, &dwReaders)
				== SCARD_S_SUCCESS) && (dwReaders != dwReadersOld))
				goto get_readers;
		}

		/* Timestamp the event as we get notified */
		t = time(NULL);
		printf("\n%s", ctime(&t));

		/* Now we have an event, check all the readers in the list to see what
		 * happened */
		for (current_reader=0; current_reader < nbReaders; current_reader++)
		{
			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_CHANGED)
			{
				/* If something has changed the new state is now the current
				 * state */
				rgReaderStates_t[current_reader].dwCurrentState =
					rgReaderStates_t[current_reader].dwEventState;
			}
			else
				/* If nothing changed then skip to the next reader */
				continue;

			/* From here we know that the state for the current reader has
			 * changed because we did not pass through the continue statement
			 * above.
			 */

			/* Specify the current reader's number and name */
			printf("Reader %d: %s%s%s\n", current_reader,
				magenta, rgReaderStates_t[current_reader].szReader,
				color_end);

			/* Dump the full current state */
			printf("  Card state: %s", red);

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_IGNORE)
				printf("Ignore this reader, ");

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_UNKNOWN)
			{
				printf("Reader unknown\n");
				goto get_readers;
			}

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_UNAVAILABLE)
				printf("Status unavailable, ");

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_EMPTY) 
                          {
				printf("Card removed, ");
                                /* if (logged_in) { */
                                /*   printf("Logging out"); */
                                /*   system("curl http://localhost:8081/user/intergalacticSheepherder/logout"); */
                                /*   logged_in = FALSE; */
                                /* } */
                          }
                        

			if (rgReaderStates_t[current_reader].dwEventState &
                            SCARD_STATE_PRESENT) {
                          if (rgReaderStates_t[current_reader].cbAtr > 0)
                            {
                              printf("  ATR: ");
                                
                              if (rgReaderStates_t[current_reader].cbAtr)
				{
                                  for (i=0; i<rgReaderStates_t[current_reader].cbAtr; i++) {
                                    sprintf(&atr[i*3], "%02X ",
                                            rgReaderStates_t[current_reader].rgbAtr[i]);
                                    sprintf(&atr_packed[i*2], "%02X",
                                            rgReaderStates_t[current_reader].rgbAtr[i]);

                                  }
                                  atr[i*3-1] = '\0';
                                  atr_packed[i*2] = '\0';
				}
                              else
                                {
                                  atr[0] = '\0';
                                  atr_packed[0] = '\0';
                                }
                              printf("Card inserted.  "); 
                              printf("%s%s%s\n", magenta, atr, color_end);
                                
                              int count = 0;
                              int comp = 1;
                              for (count = 0; count < num_gpii_tokens; count++ ) {
                                // sprintf("The comp: %d", strncmp(atr,gpii_atrs[count],4));

                                comp = strcmp(atr,gpii_atrs[count]);
                                /* printf("The comp: %d\n", comp); */
                                /* printf(" %s\n", gpii_atrs[count]); */
                                /* printf("The 2ompstr %s\n", atr); */
                                /* printf("What the %d\n", count); */
                                if (comp == 0) {
                                  time_t current_time = time(NULL);
                                  if (logged_in) {
                                    time_t session_duration = current_time-logged_in_time;
                                    printf("Session Duration: %i", (int)session_duration);
                                    if (session_duration > 5) {
                                      char logout_cmd[1024];
                                      sprintf(&logout_cmd, "curl http://localhost:8081/user/%s/logout", gpii_tokens[count]);
                                      system(logout_cmd);
                                      printf("\n");
                                      logged_in = FALSE;
                                    }
                                  }
                                  else {           
                                    char login_cmd[1024];
                                    sprintf(&login_cmd, "curl http://localhost:8081/user/%s/login", gpii_tokens[count]);
                                    system(login_cmd);
                                    printf("\n");
                                    logged_in = TRUE;
                                    logged_in_time = time(NULL);
                                  }
                                }
                                comp = 1;  
                              }
                            }
                        }

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_ATRMATCH)
				printf("ATR matches card, ");

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_EXCLUSIVE)
				printf("Exclusive Mode, ");

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_INUSE)
				printf("Shared Mode, ");

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_MUTE)
				printf("Unresponsive card, ");

			printf("%s\n", color_end);

			/* force display */
			fflush(stdout);

		} /* for */

		rv = SCardGetStatusChange(hContext, timeout, rgReaderStates_t,
			nbReaders);
	} /* while */

	/* If we get out the loop, GetStatusChange() was unsuccessful */
	test_rv("SCardGetStatusChange", rv, hContext);

	/* We try to leave things as clean as possible */
	rv = SCardReleaseContext(hContext);
	test_rv("SCardReleaseContext", rv, hContext);

	/* free memory possibly allocated */
	if (NULL != readers)
		free(readers);
	if (NULL != rgReaderStates_t)
		free(rgReaderStates_t);

	return 0;
} /* main */

